### PR TITLE
Prepatory work for batching events to send

### PR DESCRIPTION
**Pull request by H-Shay**
This PR begins work on batching up events during the creation of a room. The PR splits out the creation and sending/persisting of the events. The first three events in the creation of the room-creating the room, joining the creator to the room, and the power levels event are sent sequentially, while the subsequent events are created and collected to be sent at the end of the function. This is currently done by appending them to a list and then iterating over the list to send, the next step (after this PR) would be to send and persist the collected events as a batch.

### CODE REVIEW

File: synapse/handlers/room.py [hunk 1]
```diff
@@ -716,7 +716,7 @@ async def create_room(
 
         if (
             self._server_notices_mxid is not None
-            and requester.user.to_string() == self._server_notices_mxid
```
Comment (H-Shay) [inline synapse/handlers/room.py:719]: This just some drive-by cleanup.

File: tests/test_federation.py [hunk 2]
```diff
@@ -46,10 +46,9 @@ def setUp(self):
         user_id = UserID("us", "test")
         our_user = create_requester(user_id)
         room_creator = self.homeserver.get_room_creation_handler()
+        config = {"preset": "public_chat"}
         self.room_id = self.get_success(
-            room_creator.create_room(
```
Comment (H-Shay) [inline tests/test_federation.py:50]: This code was actually broken for perhaps a very long time, the public chat present was never being applied but since the auth check was monkeypatched this wasn't caught, but the changes I made exposed the bug so I fixed it.

**Comment by erikjohnston** [comment]
> This change does change the structure of the dag after creating a room, with the events created after the power levels event all referencing that event as their previous event, rather than all the events in room creation sequentially referring to each other.

While this is a valid change and should work, I'm a bit cautious of us doing it like that. I would rather instead we kept the linear structure, I think this can be achieved via something like:

1. Create the event dicts
2. Convert each event dict into an actual event, using the previous events to fill in `prev_events` and `auth_events`.
3. For now I'd then just call `handle_new_client_event` in a loop, and try and batch those up in a separate PR.

Happy to pair a bit on this, as it might be quite fiddly.

**Comment by squahtx** [comment]
> > This change does change the structure of the dag after creating a room, with the events created after the power levels event all referencing that event as their previous event, rather than all the events in room creation sequentially referring to each other.
> 
> While this is a valid change and should work, I'm a bit cautious of us doing it like that. I would rather instead we kept the linear structure, I think this can be achieved via something like:
> 
> 1. Create the event dicts
> 2. Convert each event dict into an actual event, using the previous events to fill in `prev_events` and `auth_events`.
> 3. For now I'd then just call `handle_new_client_event` in a loop, and try and batch those up in a separate PR.
> 
> Happy to pair a bit on this, as it might be quite fiddly.

NB: batch sending of linear chains of events is #12010, which @reivilibre proposed a while back.

Also we need to take care regarding `EventContext`s - they contain the state groups before and after each event, which should include the previous, unsent events in the batch.

File: synapse/handlers/message.py [hunk 3]
```diff
@@ -554,6 +554,98 @@ def __init__(self, hs: "HomeServer"):
...
+        )
+
+        # Pass on the outlier property from the builder to the event
+        # after it is created
+        if builder.internal_metadata.outlier:
+            event.internal_metadata.outlier = True
+
+        self.validator.validate_new(event, self.config)
+        await self._validate_event_relation(event)
+        logger.debug("Created event %s", event.event_id)
```
Comment (erikjohnston) [inline synapse/handlers/message.py:645]: Can we call this code from the existing method to avoid duplication?

File: synapse/handlers/room.py [hunk 4]
```diff
@@ -1040,46 +1041,65 @@ async def _send_events_for_new_room(
...
 
         creator_id = creator.user.to_string()
-
         event_keys = {"room_id": room_id, "sender": creator_id, "state_key": ""}
-
         depth = 1
         last_sent_event_id: Optional[str] = None
+        prev_event: List[str] = []
+        state_map = {}
+        auth_events = []
```
Comment (erikjohnston) [inline synapse/handlers/room.py:1049]: Can you add some comments for these vars?

File: synapse/handlers/room.py [hunk 5]
```diff
@@ -1040,46 +1041,65 @@ async def _send_events_for_new_room(
...
-            ) = await self.event_creation_handler.create_and_send_nonmember_event(
+            event_dict = create_event_dict(etype, content, **kwargs)
+
+            event = await self.event_creation_handler.create_event_for_batch(
                 creator,
-                event,
+                event_dict,
+                prev_event,
+                depth=depth,
+                auth_event_ids=auth_event_ids,
```
Comment (erikjohnston) [inline synapse/handlers/room.py:1075]: I think we want to derive this from `state_map`? We can use `hs.get_event_auth_handler().compute_auth_events(..)` to go from state_map to auth_event_ids

Comment (H-Shay) [reply synapse/handlers/room.py:1075]: I am slightly confused by this-`compute_auth_events` takes an already-created event and calculates what type of events to pull out of the full state (the `state_map`) to auth that particular event, but in this case we need the auth events to create the event-I think I am missing something here?
My solution was to just collect and pass in the state events I knew would be needed for authorizing all the subsequent events (thus the create, membership and power level event ids become the `auth_event_ids` as they are created), but this might not be right?

Comment (erikjohnston) [reply synapse/handlers/room.py:1075]: > takes an already-created event

You can give it an `EventBuilder` object, which doesn't need to have been finalised yet? You can get one from:

```
builder = self.event_builder_factory.for_room_version(
            room_version_obj, event_dict
        )
```

(You'd probably then want to change the new create_event method to take in `builder` rather than `event_dict`)

File: synapse/handlers/message.py (hunk 3 above)
Comment (H-Shay) [reply synapse/handlers/message.py]: So I tried to use `create_event` for both cases (batched and non-batched) rather than having two separate functions but ran into in issue with the fact that we return a tuple of the event and event context in the case of non-batched events but only the event in the case of batched events. This requires a change to the return signature, and changing it to `Union[EventBase, Tuple[EventBase, EventContext]]` resulted in a cascade of mypy errors. I thought it cleaner to split the `create_event` back out into two functions and to add some helper methods to encapsulate some of the duplicated code rather than adding a billion `assert foo is not blah` everywhere, but there are still some duplicated lines causing sonarcube to be unhappy. Is there a more graceful way to handle this?

File: synapse/handlers/room.py (hunk 5 above)
Comment (H-Shay) [reply synapse/handlers/room.py]: Right this has been done.

File: synapse/handlers/message.py [hunk 6]
```diff
@@ -735,6 +697,126 @@ async def create_event(
 
         return event, context
 
+    async def create_event_for_batch(
```
Comment (erikjohnston) [inline synapse/handlers/message.py:700]: Can we call this from within `create_event`, rather than duplicating bits of the code? I see that we've split out some of the common functionality, which is great, but it might just be easier to call one from the other instead?

File: synapse/state/__init__.py [hunk 7]
```diff
@@ -420,6 +420,69 @@ async def compute_event_context(
             partial_state=partial_state,
         )
 
+    async def compute_event_context_for_batched(
+        self,
+        event: EventBase,
+        state_ids_before_event: StateMap[str],
+        current_state_group: int,
+    ) -> EventContext:
```
Comment (erikjohnston) [inline synapse/state/__init__.py:428]: Is this now basically the same as `state_ids_before_event`, except with an extra `current_state_group`?

**Review by erikjohnston — COMMENTED** [review]
I think functionally this looks good! Just wondering if we can simplify things a bit.

File: synapse/state/__init__.py (hunk 7 above)
Comment (erikjohnston) [reply synapse/state/__init__.py:428]: Actually, I guess we'll want to keep these functions different as we'll be changing this in future iterations?

File: synapse/handlers/message.py [hunk 8]
```diff
@@ -567,9 +567,17 @@ async def create_event(
         outlier: bool = False,
         historical: bool = False,
         depth: Optional[int] = None,
+        state_map: Optional[Mapping[Tuple[str, str], str]] = None,
```
Comment (erikjohnston) [inline synapse/handlers/message.py:570]:
```suggestion
        state_map: Optional[StateMap[str]] = None,
```

File: synapse/handlers/room.py [hunk 9]
```diff
@@ -1053,46 +1055,74 @@ async def _send_events_for_new_room(
...
+        # the last event sent/persisted to the db
         last_sent_event_id: Optional[str] = None
-
-        def create(etype: str, content: JsonDict, **kwargs: Any) -> JsonDict:
+        # the most recently created event
+        prev_event: List[str] = []
+        # a map of event types, state keys -> event_ids. We collect these mappings this as events are
+        # created (but not persisted to the db) to determine state for future created events
+        # (as this info can't be pulled from the db)
+        state_map: dict = {}
```
Comment (erikjohnston) [inline synapse/handlers/room.py:1067]:
```suggestion
        state_map: MutableStateMap[str] = {}
```

File: synapse/handlers/message.py [hunk 10]
```diff
@@ -685,23 +704,40 @@ async def create_event(
...
         # this behaviour isn't spec'd yet, meaning we should be able to deactivate this
         # behaviour. Another reason is that this code is also evaluated each time a new
         # m.room.aliases event is created, which includes hitting a /directory route.
         # Therefore not including this condition here would render the similar one in
         # synapse.handlers.directory pointless.
-        if builder.type == EventTypes.Aliases and self.require_membership_for_aliases:
+        if (
+            builder.type == EventTypes.Aliases
+            and self.require_membership_for_aliases
+            and not for_batch
```
Comment (erikjohnston) [inline synapse/handlers/message.py:739]: Why do we not do this for `for_batch`?

File: synapse/state/__init__.py [hunk 11]
```diff
@@ -282,7 +282,6 @@ async def compute_event_context(
             RuntimeError if `state_ids_before_event` is not provided and one or more
                 prev events are missing or outliers.
         """
-
```
Comment (erikjohnston) [inline synapse/state/__init__.py:285]: Can we remove these unrelated changes from the PR please?

File: synapse/handlers/message.py [hunk 12]
```diff
@@ -685,23 +704,40 @@ async def create_event(
...
+            assert prev_event_ids is not None
+            assert state_map is not None
+            assert current_state_group is not None
+            auth_ids = self._event_auth_handler.compute_auth_events(builder, state_map)
+            event = await builder.build(
+                prev_event_ids=prev_event_ids, auth_event_ids=auth_ids, depth=depth
+            )
+            context = await self.state.compute_event_context_for_batched(
+                event, state_map, current_state_group
+            )
```
Comment (erikjohnston) [inline synapse/handlers/message.py:717]: I think we can move this into `create_new_client_event`? That way we can always call `create_new_client_event`, and so we don't have to worry about if we've done everything we need to, e.g. atm this code path isn't calling `third_party_event_rules.check_event_allowed` (which may or may not be an issue)?

File: synapse/handlers/message.py (hunk 10 above)
Comment (H-Shay) [reply synapse/handlers/message.py:739]: My (admittedly poor) thinking on this was that since `for_batch` is only used with the createRoom codepath we already know that the sender is a member, but your question has made me realize this won't necessarily hold true if other code paths employ this mechanism. I have changed it so that the code now looks for the membership event id in the state map before trying to pull it from the DB if `for_batch` is true.

**Review by erikjohnston — APPROVED** [review]
Woo!
